### PR TITLE
[Feature][Torchax] Add support for Qwen3-Omni-30B-A3B

### LINKED ISSUE #2712: [Feature]: Qwen3-Omni-30B-A3B
### 🚀 The feature, motivation and pitch

Add support for Qwen3-Omni-30B-A3B from the torchax path 

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://github.com/vllm-project/tpu-inference/tree/main/docs), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by amanseervi**
# Description
This PR adds support for the Qwen3-Omni-30B-A3B model via the TorchAX path. Modalities supported are:
 - Image
 - Video
 - Audio

The implementation is done via the Torchax route, so for the most part, it runs with the upstream vLLM implementation. The exception is a few specific functions where patches were made to avoid JAX-unfriendly patterns to make them safe for JIT-compilation on TPUs.

### Files changed/added :
*   **`tpu_inference/models/vllm/experimental/qwen3_omni_patcher.py`**: 
    Added Qwen3-Omni specific patches for stateless Deepstack support (imported from Qwen3-VL patches), JIT-friendly vision/ audio tower forward pass. Replaced dynamic boolean indexing with simple static slicing loops to keep shapes constant. Added a global patch for `torch.nn.utils.rnn.pad_sequence` using pure-PyTorch slicing to bypass a silent precision/tracing corruption bug caused by PyTorch's native C++ padding op on TPUs.
*   **`tpu_inference/models/vllm/vllm_model_wrapper.py`**: 
    Updated `wrap_embed_multimodal_func` to pop `audio_feature_lengths` as a static top-level argument to support the audio modality.
*   **`tpu_inference/models/vllm/experimental/vision_tower_jit.py`**: 
    Updated `maybe_prepare_for_jit` to recursively convert raw audio feature length tensors into static Python `tuple`s of plain `int`s, and added `"audio_feature_lengths"` to `static_argnames` under `maybe_jit_embed_multimodal_func` so JAX JIT treats them as compile-time constants.
* **`tpu_inference/models/vllm/experimental/qwen3_vl_patcher.py`**: 
  Updated `_patched_set_deepstack` to allow deepstack_embeds as single tensor input, since upstream Omni code does it that way.

FIXES: #2712

# Tests

Server : 
```bash
vllm serve "Qwen/Qwen3-Omni-30B-A3B-Instruct" \
    --tensor-parallel-size 4 \
    --max-model-len 5000 \
    --max-num-batched-tokens 8192 \
    --trust-remote-code \
    --gpu-memory-utilization 0.85 \
    --port 8000 \
    --limit-mm-per-prompt '{"image": 1,"video": 1,"audio": 1}' \
    --mm-processor-kwargs '{"max_pixels": 3000000}' \
    --no-enable-prefix-caching
```

Image : 
```bash
curl http://localhost:8000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "Qwen/Qwen3-Omni-30B-A3B-Instruct",
        "messages": [
            {
                "role": "user",
                "content": [
                    { "type": "text", "text": "Hello, please introduce this picture" },
                    { "type": "image_url", "image_url": { "url": "https://a-z-animals.com/media/swan-3.jpg" } }
                ]
            }
        ]
    }'
```

```bash
{"id":"chatcmpl-805921b2000d887d","object":"chat.completion","created":1779730396,"model":"Qwen/Qwen3-Omni-30B-A3B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"Of course, here is a description of the image provided.\n\nThe image is a beautiful, clear photograph of a white swan gracefully floating on the surface of calm, blue water.\n\n- **Subject**: The central figure is a swan, characterized by its pristine white plumage, a long, elegant, and gracefully curved S-shaped neck, and a sleek body.\n- **Features**: It possesses a prominent, bright orange-red beak, which is accented with a noticeable black patch, known as the basal knob, at the base where the beak meets the forehead. The swan's head is slightly turned, and its dark eye is just visible, giving it a calm and regal expression.\n- **Posture and Environment**: The swan is resting, with its wings held close to its body and slightly raised, which emphasizes the volume and texture of its feathers. It is calmly floating on water that appears to be a lake or a large pond. The surface of the water is relatively still, showing gentle ripples and reflecting the swan.\n- **Reflection**: A clear, though slightly rippled, reflection of the swan can be seen in the water directly beneath it. This echo-like detail adds depth and symmetry to the picturesque scene.\n- **Overall Impression**: The lighting is bright and soft, suggesting it's taken during the day. The overall mood is one of serenity, peace, and natural beauty, with the contrast between the white swan and the blue water making the image visually striking.","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null},"logprobs":null,"finish_reason":"stop","stop_reason":null,"token_ids":null,"routed_experts":null}],"service_tier":null,"system_fingerprint":"vllm-0.21.1rc1.dev159+g9640970de-tp4-16d304ef","usage":{"prompt_tokens":667,"total_tokens":974,"completion_tokens":307,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"prompt_text":null,"kv_transfer_params":null}
```

Video : 

```bash
curl http://localhost:8000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "Qwen/Qwen3-Omni-30B-A3B-Instruct",
        "messages": [
            {
                "role": "user",
                "content": [
                    { "type": "text", "text": "Describe what happens in this video." },
                    { "type": "video_url", "video_url": { "url": "https://github.com/intel-iot-devkit/sample-videos/raw/master/bottle-detection.mp4" } }
                ]
            }
        ]
    }'
```

```bash
{"id":"chatcmpl-bdf0ec1372930ba4","object":"chat.completion","created":1779730481,"model":"Qwen/Qwen3-Omni-30B-A3B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"Here is a detailed, step-by-step description of what happens in the video, based on the provided frames:\n\nThe video takes place on a flat, white surface in front of a plain white backdrop.\n\n**Initial State:**\nA hand sequentially removes three identical empty clear plastic bottles that are lined up horizontally on the surface.\n- First, the hand appears from the left and removes the bottle positioned at the far left.\n- Next, the hand removes the bottle in the center.\n- Finally, the hand removes the bottle at the far right.\nThe surface is now empty, and the hand leaves the frame.\n\n**Span of Time:**\nFor a short period, the scene remains static with just the empty white surface visible.\n\n**Restoration:**\nThe hand re-enters the frame and puts the three bottles back in the exact same lineup as before.\n\n**Repetition:**\nThe hand then reverses the previous action, removing the three bottles one by one from the scene in the same order (left, center, right), and again, the surface is left empty.","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null},"logprobs":null,"finish_reason":"stop","stop_reason":null,"token_ids":null,"routed_experts":null}],"service_tier":null,"system_fingerprint":"vllm-0.21.1rc1.dev159+g9640970de-tp4-16d304ef","usage":{"prompt_tokens":3537,"total_tokens":3753,"completion_tokens":216,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"prompt_text":null,"kv_transfer_params":null}
```

Audio : 

Install these first : 
```bash
# OS level audio encdoder
sudo apt-get install -y libsndfile1

# In the python venv
uv pip install librosa soundfile
```

```bash
curl http://localhost:8000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "Qwen/Qwen3-Omni-30B-A3B-Instruct",
        "messages": [
            {
                "role": "user",
                "content": [
                    { "type": "audio_url", "audio_url": { "url": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen3-Omni/demo/cough.wav" } },
                    { "type": "text", "text": "What can you see and hear? Answer in one short sentence." }
                ]
            }
        ]
    }'
```

```bash
{"id":"chatcmpl-8bbfdcbd8e194407","object":"chat.completion","created":1779730564,"model":"Qwen/Qwen3-Omni-30B-A3B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"The audio features the sound of a person coughing.","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null},"logprobs":null,"finish_reason":"stop","stop_reason":null,"token_ids":null,"routed_experts":null}],"service_tier":null,"system_fingerprint":"vllm-0.21.1rc1.dev159+g9640970de-tp4-16d304ef","usage":{"prompt_tokens":61,"total_tokens":73,"completion_tokens":12,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"prompt_text":null,"kv_transfer_params":null}
```

### Run Summary
* **Total Samples/Requests:** 900
* **Successful API Calls:** 897 (3 failed due to exceeding maximum context length constraints)
* **Total Token Count:** 520,324 tokens (Input: 512,486 | Output: 7,838)

### Final Aggregated Summary

| Task | Filter | n-shot | Metric | Value |
| :--- | :--- | :---: | :--- | :---: |
| **mmmu_val** | none | 0 | mmmu_acc (↑) | **0.5933** |

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/models/vllm/experimental/vision_tower_jit.py [hunk 1]
```diff
@@ -54,6 +58,19 @@ def has_jittable_vision(vllm_model) -> bool:
         is_qwen3_vl
     return is_jittable_architecture(vllm_model) or is_qwen3_vl(vllm_model)
 
+def get_vision_config(hf_config: Any) -> Any:
+    """Extract vision configuration from hf_config, supporting nested/thinker wrappers."""
+
+    if hasattr(hf_config, "vision_config"):
+        return hf_config.vision_config
+    # Dynamic inspection of attributes (e.g. thinker_config)
```
Comment (lk-chen) [inline tpu_inference/models/vllm/experimental/vision_tower_jit.py]: I honestly feel it's more readable with
```
if hasattr(hf_config, "vision_config"):
    return hf_config.vision_config
if isinstance(hf_config, ....):
    return hf_config.thinker_config.vision_config

File: tpu_inference/models/vllm/vllm_model_wrapper.py [hunk 2]
```diff
@@ -466,6 +469,10 @@ def embed_multimodal_func_torch(params_and_buffers: Any,
 
                 kwargs = maybe_prepare_for_jit(kwargs, self.model.vllm_model)
 
+                # Extract audio_feature_lengths from the top-level kwargs so it is not moved to JAX
+                audio_feature_lengths = kwargs.pop("audio_feature_lengths",
```
Comment (lk-chen) [inline tpu_inference/models/vllm/vllm_model_wrapper.py]: is it necessary? In `maybe_prepare_for_jit` you already turned audio_feat_lengths into tuple, so `move` should have no op

File: tpu_inference/models/vllm/experimental/vision_tower_jit.py [hunk 3]
```diff
@@ -183,4 +201,15 @@ def maybe_prepare_for_jit(kwargs: dict, vllm_model) -> dict:
     for k, v in kwargs.items():
         if k in ("image_grid_thw", "video_grid_thw", "grid_thw"):
             kwargs[k] = GridTHW(v.tolist())
-    return kwargs
+
+    def _convert(obj, key=None):
```
Comment (lk-chen) [inline tpu_inference/models/vllm/experimental/vision_tower_jit.py]: `audio_feature_lengths` is NOT nested, right? so it could be simpler to 
```
if k in ("audio_feature_lengths", ):
    kwargs[k] = tuple(.....tolist())
```
at line 204, and avoid `_convert`

If that's not the case, feel free to explain how `audio_feature_lengths` is nested in the input

File: tpu_inference/models/vllm/vllm_model_wrapper.py (hunk 2 above)
Comment (amanseervi) [reply tpu_inference/models/vllm/vllm_model_wrapper.py]: It isn't necessary, just did to help avoid the warning message : 
`WARNING 05-27 08:43:30 [vllm_model_wrapper.py:478] Expect torch.Tensor, got <class 'int'>`.

We get similar warning logs for GridTHW when processing images, since all static args are tuples. 
I have removed it for now. We can maybe skip the `move` call for the static args in some later PR if we do want to suppress the warnings.

File: tpu_inference/models/vllm/experimental/vision_tower_jit.py (hunk 1 above)
Comment (amanseervi) [reply tpu_inference/models/vllm/experimental/vision_tower_jit.py]: Done

File: tpu_inference/models/vllm/experimental/vision_tower_jit.py (hunk 3 above)
Comment (amanseervi) [reply tpu_inference/models/vllm/experimental/vision_tower_jit.py]: Yes, it is directly available in the top level `kwargs`.  Simplified it now.

File: tpu_inference/models/vllm/vllm_model_wrapper.py [hunk 4]
```diff
@@ -264,6 +264,10 @@ def load_weights(self,
         # NOTE: Apply Qwen3-VL model specific patches
         maybe_apply_qwen3_vl_patches(self.model.vllm_model)
 
+        from tpu_inference.models.vllm.experimental.qwen3_omni_patcher import \
```
Comment (lk-chen) [inline tpu_inference/models/vllm/vllm_model_wrapper.py]: qwen3_vl is importing at top, please follow the pattern

File: tpu_inference/models/vllm/experimental/vision_tower_jit.py [hunk 5]
```diff
@@ -55,6 +59,16 @@ def has_jittable_vision(vllm_model) -> bool:
     return is_jittable_architecture(vllm_model) or is_qwen3_vl(vllm_model)
 
 
+def get_vision_config(hf_config: Any) -> Any:
+    """Extract vision configuration from hf_config, supporting nested/thinker wrappers."""
+
+    if hasattr(hf_config, "vision_config"):
+        return hf_config.vision_config
+    if hasattr(hf_config, "thinker_config"):
```
Comment (lk-chen) [inline tpu_inference/models/vllm/experimental/vision_tower_jit.py]: prefer to check instance type instead of attribute existance

File: tpu_inference/models/vllm/vllm_model_wrapper.py [hunk 6]
```diff
@@ -466,6 +469,10 @@ def embed_multimodal_func_torch(params_and_buffers: Any,
 
                 kwargs = maybe_prepare_for_jit(kwargs, self.model.vllm_model)
 
+                # Extract audio_feature_lengths from the top-level kwargs so it is not moved to JAX
+                # audio_feature_lengths = kwargs.pop("audio_feature_lengths",
```
Comment (lk-chen) [inline tpu_inference/models/vllm/vllm_model_wrapper.py]: clean up, and line 489

File: tpu_inference/models/vllm/experimental/vision_tower_jit.py (hunk 5 above)
Comment (amanseervi) [reply tpu_inference/models/vllm/experimental/vision_tower_jit.py]: Do you mean something like this : 
```python
if isinstance(hf_config, Qwen3OmniMoeConfig):
            return hf_config.thinker_config.vision_config
```

Would't that be too model specific?

File: tpu_inference/models/vllm/vllm_model_wrapper.py (hunk 4 above)
Comment (amanseervi) [reply tpu_inference/models/vllm/vllm_model_wrapper.py]: Done

File: tpu_inference/models/vllm/vllm_model_wrapper.py (hunk 6 above)
Comment (amanseervi) [reply tpu_inference/models/vllm/vllm_model_wrapper.py]: Done

File: tpu_inference/models/vllm/experimental/vision_tower_jit.py (hunk 5 above)
Comment (lk-chen) [reply tpu_inference/models/vllm/experimental/vision_tower_jit.py]: right. The `JITTABLE_MODEL_ARCH` is already model specific

Comment (amanseervi) [reply tpu_inference/models/vllm/experimental/vision_tower_jit.py]: Done. Replaced the entire thing with this now : 
```python
if isinstance(hf_config, Qwen3OmniMoeConfig):
        return hf_config.thinker_config.vision_config
return hf_config.vision_config
```

File: tpu_inference/models/vllm/experimental/qwen3_vl_patcher.py
Comment (lk-chen) [inline tpu_inference/models/vllm/experimental/qwen3_vl_patcher.py:1]: Is this file change also necessary? "Files changed/added" section doesn't mention this file

**Review by lk-chen — APPROVED** [review]
LGTM with question on qwen3_vl_patcher

File: tpu_inference/models/vllm/experimental/qwen3_vl_patcher.py
Comment (amanseervi) [reply tpu_inference/models/vllm/experimental/qwen3_vl_patcher.py:1]: Yes this is required.  Omni patcher reuses `_patched_set_deepstack` from the VL patcher. The VL patcher passes a dict to the setter, Omni passes a raw 3D Tensor directly (matching upstream). Since the original VL setter only handled dicts/lists, we had to add this tensor case. I'll update the description.

File: tpu_inference/models/vllm/vllm_model_wrapper.py
Comment (kyuyeunk) [inline tpu_inference/models/vllm/vllm_model_wrapper.py:1]: i feel like all the "patches" need to be consolidated into a single function - instead of `vllm_model_wrapper.py` calling individual patchers of each models.

**Review by kyuyeunk — APPROVED** [review]
will approve, but if you can resolve / work on the comment i've left, that would be great.

File: tpu_inference/models/vllm/vllm_model_wrapper.py
Comment (amanseervi) [reply tpu_inference/models/vllm/vllm_model_wrapper.py:1]: Something like `vllm_model_wrapper` calling a generic `maybe_apply_patches`, which will then call specific patch based on model, is it?

Comment (amanseervi) [reply tpu_inference/models/vllm/vllm_model_wrapper.py:1]: I felt `model_patcher.py` would be best suited for such a function. Added `apply_model_specific_patches` there which does the model specific patching.
